daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala

```diff
@@ -0,0 +1,197 @@
+      val original = lam(2, app(arg0, app(arg0, app(arg0, arg1))))
+      val expected =
+        AExpr(lam(2, let1(appa(arg0, arg1), let1(appa(arg0, stack1), appa(arg0, stack1)))))
+      testTransform(original, expected)
+    }
+  }
+
+  "arithmetic non-atomic: [\\f x. f (x+1)]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(2, app(arg0, binop(SBAddInt64, arg1, num1)))
+      val expected = AExpr(lam(2, let1b2(SBAddInt64, arg1, num1, appa(arg0, stack1))))
+      testTransform(original, expected)
+    }
+  }
+
+  "nested (4x non-atomic): [\\f x. f(x+1) - f(x+2)]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(
+          2,
+          binop(
+            SBSubInt64,
+            app(arg0, binop(SBAddInt64, arg1, num1)),
+            app(arg0, binop(SBAddInt64, arg1, num2))))
+      val expected =
+        AExpr(
+          lam(
+            2,
+            let1b2(
+              SBAddInt64,
+              arg1,
+              num1,
+              let1(
+                appa(arg0, stack1),
+                let1b2(
+                  SBAddInt64,
+                  arg1,
+                  num2,
+                  let1(appa(arg0, stack1), binopa(SBSubInt64, stack3, stack1)))))))
+      testTransform(original, expected)
+    }
+  }
+
+  "multi-arg fun: [\\f g. f (g 1) (g 2)]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(2, app2(arg0, app(arg1, num1), app(arg1, num2)))
+      val expected =
+        AExpr(lam(2, let1(appa(arg1, num1), let1(appa(arg1, num2), app2a(arg0, stack2, stack1)))))
+      testTransform(original, expected)
+    }
+  }
+
+  "case expression: [\\a b c. if a then b else c]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(3, ite(arg0, arg1, arg2))
+      val expected = AExpr(lam(3, itea(arg0, arg1, arg2)))
+      testTransform(original, expected)
+    }
+  }
+
+  "non-atomic in branch: [\\f x. if x==0 then 1 else f (div(1,x))]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(2, ite(binop(SBEqual, arg1, num0), num1, app(arg0, binop(SBDivInt64, num1, arg1))))
+      val expected =
+        AExpr(
+          lam(
+            2,
+            let1b2(
+              SBEqual,
+              arg1,
+              num0,
+              itea(stack1, num1, let1b2(SBDivInt64, num1, arg1, appa(arg0, stack1))))))
+      testTransform(original, expected)
+    }
+  }
+
+  "nested lambda: [\\f g x. g (\\y. f (f y)) x]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(3, app2(arg1, clo1(arg0, 1, app(free0, app(free0, arg0))), arg2))
+      val expected =
+        AExpr(
+          lam(
+            3,
+            let1(
+              clo1(arg0, 1, let1(appa(free0, arg0), appa(free0, stack1))),
+              app2a(arg1, stack1, arg2))))
+      testTransform(original, expected)
+    }
+  }
```

Comment:
Can we please have a test case for the expression mentioned in https://github.com/digital-asset/daml/issues/6535 as well?

Reply:
done. turns out we never did add redundant lets in this example